// module-app-micro-base/web-dev-server.js
/*************************************************************************
 *
 * COMPRO CONFIDENTIAL
 * __________________
 *
 *  [2015] - [2020] Compro Technologies Private Limited
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Compro Technologies Private Limited. The
 * intellectual and technical concepts contained herein are
 * proprietary to Compro Technologies Private Limited and may
 * be covered by U.S. and Foreign Patents, patents in process,
 * and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Compro Technologies Pvt. Ltd..
 ***************************************************************************/
 
/********************************************************************
 * Web Dev server setup
 ********************************************************************/
 'use strict';

const Webpack = require('webpack');
const WebpackDevServer = require('webpack-dev-server');

module.exports = function (webDevServerOptions) {

  // First we fire up Webpack and pass in the configuration we
  // created
  var bundleStart = null;
  var compiler = Webpack(webDevServerOptions.webpackConfig);

  // We give notice in the terminal when it starts bundling and
  // set the time it started
  compiler.plugin('compile', function() {
    syslog.info('Bundling...');
    bundleStart = Date.now();
  });

  // We also give notice when it is done compiling, including the
  // time it took.
  compiler.plugin('done', function() {
     syslog.info('Bundled in ' + (Date.now() - bundleStart) + 'ms!');
  });

  var bundler = new WebpackDevServer(compiler, {

    publicPath: '/',
    
    // To serve your index.html in place of 404 responses
    historyApiFallback: {
      disableDotRule: true
    },

    // Configure hot replacement
    hot: true,

    // The rest is terminal configurations
    quiet: false, //output anything to the console.
    noInfo: true, //suppress useless information
    stats: {
      colors: true //add some colors to the output
    },
    proxy: { // All API request are proxied to Express-server running on 3000 port
      '/apigateway/*': {
          target: 'http://localhost:' + webDevServerOptions.expressServerPort  + '/'
        }
     }
  });

  // We fire up the development server and give notice in the terminal
  // that we are starting the initial bundle
  bundler.listen(webDevServerOptions.webDevServerPort, 'localhost', function () {
    syslog.info('Web-dev-server running on port ' + webDevServerOptions.webDevServerPort);
  });

};
