// module-app-micro-base/libs/app-constants.js
/*************************************************************************
 *
 * COMPRO CONFIDENTIAL
 * __________________
 *
 *  [2015] - [2020] Compro Technologies Private Limited
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Compro Technologies Private Limited. The
 * intellectual and technical concepts contained herein are
 * proprietary to Compro Technologies Private Limited and may
 * be covered by U.S. and Foreign Patents, patents in process,
 * and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Compro Technologies Pvt. Ltd..
 ***************************************************************************/

/********************************************************************
 * Application Constants 
 ********************************************************************/
'use strict';

const CONSTANTS = {
    APP: {
        ERROR_TYPES: {
            DLS_CORE: 'DLS_CORE_ERROR',
            NODE_SERVER: 'NODE_SERVER_ERROR',
            REDIS: 'REDIS_ERROR',
            UNCAUGHT: 'UNCAUGHT_ERROR',
            EXIT: 'PROCESS_EXIT',
            REJECTION: 'UNHANDLED_REJECTION',
            TOKEN_NOT_FOUND: 'TOKEN_NOT_FOUND_ERROR',
            INTERNAL_SERVER_ERROR: "Internal Server Error",
            INVALID_JWT: 'INVALID_JWT',
            GIGYA_ERROR: "GIGYA_ERROR",
            GIGYA_ACCOUNT_ERROR: "GIGYA_ACCOUNT_ERROR",
            PASSWORD_CHANGE_INVALID_USER: 'PASSWORD_CHANGE_INVALID_USER',
            GIGYA_DS_ERROR: "GIGYA_DS_ERROR",
            GETSTREAM_COMMENTS_ERROR: "GETSTREAM_COMMENTS_ERROR",
            GIGYA_JWK_ERROR: "GIGYA_JWK_ERROR",
            GIGYA_EDUELT_ERROR: "GIGYA_EDUELT_ERROR",
            PTAS: {
                CREATE_CLIENT: 'PTAS_CREATE_CLIENT_ERROR',
                AUTHENTICATION_API: 'PTAS_AUTHENTICATION_API_ERROR',
                USER_CREATE_API: 'PTAS_USER_CREATE_API_ERROR',
                ACTIVATE_PRODUCT_API: 'PTAS_ACTIVATE_PRODUCT_API_ERROR'
            },
            PEAS: {
                PEAS_COMMUNICATION_ERROR: 'PEAS_COMMUNICATION_ERROR',
                PEAS_WORKFLOW_ERROR: 'PEAS_WORKFLOW_ERROR'
            },
            ROLE_MISMATCH: "ROLE_MISMATCH_ERROR",
            AUTHORIZATION_ERROR: "AUTHORIZATION_ERROR",
            KONAKART: {
                KONAKART_ERROR: "KONAKART_ERROR",
            },
            EMAIL_ERROR: {
                SES_EMAIL: 'SES_EMAIL_ERROR'
            },
            LTI: {
                LAUNCH_ERROR: {
                    GIGYA_ERROR: 'LTI_LAUNCH_GIGYA_ERROR',
                    GATEWAY_TOOL_ERROR: 'LTI_LAUNCH_GATEWAY_TOOL_ERROR',
                    NODE_SERVER: 'LTI_LAUNCH_NODE_SERVER'
                },
                DEEPLINK_ERROR: {
                    GATEWAY_TOOL_ERROR: 'LTI_DEEPLINK_GATEWAY_TOOL_ERROR',
                },
                DYNAMODB_ERROR: {
                    GITRACKER: 'GITRACKER_DYNAMODB_ERROR'
                }
            },
            GETSTREAM_CONTENT_NOTIFICATION_ERROR: "GETSTREAM_CONTENT_NOTIFICATION_ERROR"
        },
        ERROR_CODES: {
            GETSTREAM_COMMENTS_ERROR: {
                COMMENT_VALIDATION_FAILED: "COMMENT_VALIDATION_FAILED",
                USER_TEXT_TOO_BIG: "USER_TEXT_TOO_BIG",
            },
            KONAKART: {
                KONAKART_EXCEPTION: "KONAKART_EXCEPTION",
                KONAKART_PRODUCT_NOT_FOUND: "KONAKART_PRODUCT_NOT_FOUND",
                KONAKART_REDIRECT_FIELDS_NOT_FOUND: "KONAKART_REDIRECT_FIELDS_NOT_FOUND",
                KONAKART_API_FAILURE: "KONAKART_API_FAILURE",
                KONAKART_PAYMENT_UNSUCCESSFULL: "KONAKART_PAYMENT_UNSUCCESSFULL",
                KONAKART_GET_PAYMENT_DETAILS_EMPTY_RESPONSE: "KONAKART_GET_PAYMENT_DETAILS_EMPTY_RESPONSE",
                KONAKART_PRODUCT_CURRENCY_NOT_FOUND: "KONAKART_PRODUCT_CURRENCY_NOT_FOUND",
                KONAKART_PRODUCT_PRICE_NOT_FOUND: "KONAKART_PRODUCT_PRICE_NOT_FOUND",
                KONAKART_CURRENCY_CODE_DOES_NOT_EXIST: "KONAKART_CURRENCY_CODE_DOES_NOT_EXIST",
                DEFAULT_COUNTRY_CODE_NOT_FOUND: "DEFAULT_COUNTRY_CODE_NOT_FOUND"
            }
        },
        INFO_TYPES: {
            PTAS_INVALID_ACTIVATION_CODE: 'PTAS_INVALID_ACTIVATION_CODE',
            PTAS_ACTIVATION_CODE_ALREADY_USED: 'PTAS_ACTIVATION_CODE_ALREADY_USED',
            PTAS_PRODUCT_NOT_RELEASED: 'PTAS_PRODUCT_NOT_RELEASED',
            PTAS_PRODUCT_TYPE_MISMATCH: 'PTAS_PRODUCT_TYPE_MISMATCH',
            PROXY_LOGIN_SUCCESS: 'PROXY_LOGIN_SUCCESS'
        },
        LAUNCH_TYPES: {
            LTI: "lti"
        },
        COUNTRY_TO_CURRENCY_MAP : {
            "defaultCurrency": "USD",
            "GB": {
              "currencyCode": "GBP"
            },
            "GG": {
              "currencyCode": "GBP"
            },
            "IM": {
              "currencyCode": "GBP"
            },
            "JE": {
              "currencyCode": "GBP"
            },
            "IE": {
              "currencyCode": "EUR"
            },
            "IO": {
              "currencyCode": "GBP"
            },
            "GS": {
              "currencyCode": "GBP"
            },
            "SH": {
              "currencyCode": "GBP"
            },
            "AL": {
              "currencyCode": "EUR"
            },
            "AM": {
              "currencyCode": "EUR"
            },
            "AZ": {
              "currencyCode": "EUR"
            },
            "BY": {
              "currencyCode": "EUR"
            },
            "BA": {
              "currencyCode": "EUR"
            },
            "BG": {
              "currencyCode": "EUR"
            },
            "HR": {
              "currencyCode": "EUR"
            },
            "CZ": {
              "currencyCode": "EUR"
            },
            "EE": {
              "currencyCode": "EUR"
            },
            "GE": {
              "currencyCode": "EUR"
            },
            "HU": {
              "currencyCode": "EUR"
            },
            "KZ": {
              "currencyCode": "EUR"
            },
            "CS": {
              "currencyCode": "EUR"
            },
            "KG": {
              "currencyCode": "EUR"
            },
            "LV": {
              "currencyCode": "EUR"
            },
            "LT": {
              "currencyCode": "EUR"
            },
            "MK": {
              "currencyCode": "EUR"
            },
            "XN": {
              "currencyCode": "EUR"
            },
            "ME": {
              "currencyCode": "EUR"
            },
            "MD": {
              "currencyCode": "EUR"
            },
            "PL": {
              "currencyCode": "EUR"
            },
            "RO": {
              "currencyCode": "EUR"
            },
            "RU": {
              "currencyCode": "EUR"
            },
            "RS": {
              "currencyCode": "EUR"
            },
            "SI": {
              "currencyCode": "EUR"
            },
            "SK": {
              "currencyCode": "EUR"
            },
            "TJ": {
              "currencyCode": "EUR"
            },
            "TM": {
              "currencyCode": "EUR"
            },
            "UA": {
              "currencyCode": "EUR"
            },
            "UZ": {
              "currencyCode": "EUR"
            },
            "DK": {
              "currencyCode": "EUR"
            },
            "FO": {
              "currencyCode": "EUR"
            },
            "GL": {
              "currencyCode": "EUR"
            },
            "FI": {
              "currencyCode": "EUR"
            },
            "AX": {
              "currencyCode": "EUR"
            },
            "IS": {
              "currencyCode": "EUR"
            },
            "NO": {
              "currencyCode": "EUR"
            },
            "SJ": {
              "currencyCode": "EUR"
            },
            "SE": {
              "currencyCode": "EUR"
            },
            "AD": {
              "currencyCode": "EUR"
            },
            "GR": {
              "currencyCode": "EUR"
            },
            "IT": {
              "currencyCode": "EUR"
            },
            "MT": {
              "currencyCode": "EUR"
            },
            "CY": {
              "currencyCode": "EUR"
            },
            "PT": {
              "currencyCode": "EUR"
            },
            "SM": {
              "currencyCode": "EUR"
            },
            "GI": {
              "currencyCode": "EUR"
            },
            "ES": {
              "currencyCode": "EUR"
            },
            "VA": {
              "currencyCode": "EUR"
            },
            "AT": {
              "currencyCode": "EUR"
            },
            "BE": {
              "currencyCode": "EUR"
            },
            "GF": {
              "currencyCode": "EUR"
            },
            "GP": {
              "currencyCode": "EUR"
            },
            "RE": {
              "currencyCode": "EUR"
            },
            "FA": {
              "currencyCode": "EUR"
            },
            "MQ": {
              "currencyCode": "EUR"
            },
            "FR": {
              "currencyCode": "EUR"
            },
            "MC": {
              "currencyCode": "EUR"
            },
            "PM": {
              "currencyCode": "EUR"
            },
            "DE": {
              "currencyCode": "EUR"
            },
            "LI": {
              "currencyCode": "EUR"
            },
            "LU": {
              "currencyCode": "EUR"
            },
            "NL": {
              "currencyCode": "EUR"
            },
            "BQ": {
              "currencyCode": "USD"
            },
            "CH": {
              "currencyCode": "EUR"
            },
            "YT": {
              "currencyCode": "EUR"
            },
            "BI": {
              "currencyCode": "USD"
            },
            "ET": {
              "currencyCode": "USD"
            },
            "KE": {
              "currencyCode": "USD"
            },
            "MG": {
              "currencyCode": "USD"
            },
            "MW": {
              "currencyCode": "USD"
            },
            "MU": {
              "currencyCode": "USD"
            },
            "RW": {
              "currencyCode": "USD"
            },
            "SC": {
              "currencyCode": "USD"
            },
            "TZ": {
              "currencyCode": "USD"
            },
            "UG": {
              "currencyCode": "USD"
            },
            "ZM": {
              "currencyCode": "USD"
            },
            "AO": {
              "currencyCode": "USD"
            },
            "BW": {
              "currencyCode": "USD"
            },
            "LS": {
              "currencyCode": "USD"
            },
            "MZ": {
              "currencyCode": "USD"
            },
            "NA": {
              "currencyCode": "USD"
            },
            "ZA": {
              "currencyCode": "USD"
            },
            "SZ": {
              "currencyCode": "USD"
            },
            "ZW": {
              "currencyCode": "USD"
            },
            "BJ": {
              "currencyCode": "USD"
            },
            "BF": {
              "currencyCode": "USD"
            },
            "CM": {
              "currencyCode": "USD"
            },
            "CF": {
              "currencyCode": "USD"
            },
            "TD": {
              "currencyCode": "USD"
            },
            "CV": {
              "currencyCode": "USD"
            },
            "GQ": {
              "currencyCode": "USD"
            },
            "GA": {
              "currencyCode": "USD"
            },
            "GW": {
              "currencyCode": "USD"
            },
            "GH": {
              "currencyCode": "USD"
            },
            "GM": {
              "currencyCode": "USD"
            },
            "GN": {
              "currencyCode": "USD"
            },
            "CI": {
              "currencyCode": "USD"
            },
            "LR": {
              "currencyCode": "USD"
            },
            "ML": {
              "currencyCode": "USD"
            },
            "NE": {
              "currencyCode": "USD"
            },
            "NG": {
              "currencyCode": "USD"
            },
            "ST": {
              "currencyCode": "USD"
            },
            "SN": {
              "currencyCode": "USD"
            },
            "SL": {
              "currencyCode": "USD"
            },
            "TG": {
              "currencyCode": "USD"
            },
            "CD": {
              "currencyCode": "USD"
            },
            "CG": {
              "currencyCode": "USD"
            },
            "DZ": {
              "currencyCode": "USD"
            },
            "BH": {
              "currencyCode": "USD"
            },
            "EG": {
              "currencyCode": "USD"
            },
            "IR": {
              "currencyCode": "USD"
            },
            "IQ": {
              "currencyCode": "USD"
            },
            "IL": {
              "currencyCode": "EUR"
            },
            "JO": {
              "currencyCode": "USD"
            },
            "KW": {
              "currencyCode": "USD"
            },
            "LB": {
              "currencyCode": "USD"
            },
            "LY": {
              "currencyCode": "USD"
            },
            "MA": {
              "currencyCode": "USD"
            },
            "DJ": {
              "currencyCode": "USD"
            },
            "ER": {
              "currencyCode": "USD"
            },
            "MR": {
              "currencyCode": "USD"
            },
            "SO": {
              "currencyCode": "USD"
            },
            "SD": {
              "currencyCode": "USD"
            },
            "OM": {
              "currencyCode": "USD"
            },
            "QA": {
              "currencyCode": "USD"
            },
            "SA": {
              "currencyCode": "USD"
            },
            "SS": {
              "currencyCode": "USD"
            },
            "SY": {
              "currencyCode": "USD"
            },
            "TN": {
              "currencyCode": "USD"
            },
            "TR": {
              "currencyCode": "EUR"
            },
            "AE": {
              "currencyCode": "USD"
            },
            "PS": {
              "currencyCode": "USD"
            },
            "YE": {
              "currencyCode": "USD"
            },
            "AI": {
              "currencyCode": "USD"
            },
            "AG": {
              "currencyCode": "USD"
            },
            "BS": {
              "currencyCode": "USD"
            },
            "BB": {
              "currencyCode": "USD"
            },
            "BM": {
              "currencyCode": "USD"
            },
            "VG": {
              "currencyCode": "USD"
            },
            "KY": {
              "currencyCode": "USD"
            },
            "CU": {
              "currencyCode": "USD"
            },
            "DM": {
              "currencyCode": "USD"
            },
            "DO": {
              "currencyCode": "USD"
            },
            "GD": {
              "currencyCode": "USD"
            },
            "HT": {
              "currencyCode": "USD"
            },
            "JM": {
              "currencyCode": "USD"
            },
            "MS": {
              "currencyCode": "USD"
            },
            "AN": {
              "currencyCode": "USD"
            },
            "MF": {
              "currencyCode": "EUR"
            },
            "AW": {
              "currencyCode": "USD"
            },
            "CW": {
              "currencyCode": "USD"
            },
            "SX": {
              "currencyCode": "EUR"
            },
            "KN": {
              "currencyCode": "USD"
            },
            "LC": {
              "currencyCode": "USD"
            },
            "VC": {
              "currencyCode": "USD"
            },
            "TT": {
              "currencyCode": "USD"
            },
            "TC": {
              "currencyCode": "USD"
            },
            "AR": {
              "currencyCode": "USD"
            },
            "BO": {
              "currencyCode": "USD"
            },
            "BR": {
              "currencyCode": "USD"
            },
            "CL": {
              "currencyCode": "USD"
            },
            "CO": {
              "currencyCode": "USD"
            },
            "EC": {
              "currencyCode": "USD"
            },
            "FK": {
              "currencyCode": "USD"
            },
            "GY": {
              "currencyCode": "USD"
            },
            "PY": {
              "currencyCode": "USD"
            },
            "PE": {
              "currencyCode": "USD"
            },
            "SR": {
              "currencyCode": "USD"
            },
            "UY": {
              "currencyCode": "USD"
            },
            "VE": {
              "currencyCode": "USD"
            },
            "AF": {
              "currencyCode": "USD"
            },
            "BD": {
              "currencyCode": "USD"
            },
            "BT": {
              "currencyCode": "USD"
            },
            "KM": {
              "currencyCode": "USD"
            },
            "IN": {
              "currencyCode": "USD"
            },
            "MV": {
              "currencyCode": "USD"
            },
            "NP": {
              "currencyCode": "USD"
            },
            "PK": {
              "currencyCode": "USD"
            },
            "LK": {
              "currencyCode": "USD"
            },
            "BN": {
              "currencyCode": "USD"
            },
            "MM": {
              "currencyCode": "USD"
            },
            "NC": {
              "currencyCode": "USD"
            },
            "KH": {
              "currencyCode": "USD"
            },
            "CC": {
              "currencyCode": "GBP"
            },
            "CN": {
              "currencyCode": "USD"
            },
            "CK": {
              "currencyCode": "GBP"
            },
            "CX": {
              "currencyCode": "GBP"
            },
            "TL": {
              "currencyCode": "USD"
            },
            "FJ": {
              "currencyCode": "USD"
            },
            "PN": {
              "currencyCode": "USD"
            },
            "PF": {
              "currencyCode": "GBP"
            },
            "HK": {
              "currencyCode": "USD"
            },
            "ID": {
              "currencyCode": "USD"
            },
            "JP": {
              "currencyCode": "USD"
            },
            "KI": {
              "currencyCode": "USD"
            },
            "LA": {
              "currencyCode": "USD"
            },
            "MO": {
              "currencyCode": "USD"
            },
            "MY": {
              "currencyCode": "USD"
            },
            "MH": {
              "currencyCode": "USD"
            },
            "MN": {
              "currencyCode": "USD"
            },
            "NR": {
              "currencyCode": "USD"
            },
            "KP": {
              "currencyCode": "USD"
            },
            "PG": {
              "currencyCode": "USD"
            },
            "PH": {
              "currencyCode": "USD"
            },
            "SG": {
              "currencyCode": "USD"
            },
            "KR": {
              "currencyCode": "USD"
            },
            "SB": {
              "currencyCode": "USD"
            },
            "TW": {
              "currencyCode": "USD"
            },
            "TH": {
              "currencyCode": "USD"
            },
            "TO": {
              "currencyCode": "USD"
            },
            "TV": {
              "currencyCode": "USD"
            },
            "VU": {
              "currencyCode": "USD"
            },
            "VN": {
              "currencyCode": "USD"
            },
            "WS": {
              "currencyCode": "USD"
            },
            "WF": {
              "currencyCode": "GBP"
            },
            "HM": {
              "currencyCode": "GBP"
            },
            "NF": {
              "currencyCode": "GBP"
            },
            "AU": {
              "currencyCode": "USD"
            },
            "BZ": {
              "currencyCode": "USD"
            },
            "CA": {
              "currencyCode": "USD"
            },
            "SV": {
              "currencyCode": "USD"
            },
            "GT": {
              "currencyCode": "USD"
            },
            "HN": {
              "currencyCode": "USD"
            },
            "MX": {
              "currencyCode": "USD"
            },
            "NI": {
              "currencyCode": "USD"
            },
            "NU": {
              "currencyCode": "GBP"
            },
            "NZ": {
              "currencyCode": "USD"
            },
            "TK": {
              "currencyCode": "GBP"
            },
            "GU": {
              "currencyCode": "USD"
            },
            "PA": {
              "currencyCode": "USD"
            },
            "CR": {
              "currencyCode": "USD"
            },
            "AS": {
              "currencyCode": "USD"
            },
            "PR": {
              "currencyCode": "USD"
            },
            "FM": {
              "currencyCode": "GBP"
            },
            "MP": {
              "currencyCode": "USD"
            },
            "PW": {
              "currencyCode": "USD"
            },
            "UM": {
              "currencyCode": "USD"
            },
            "US": {
              "currencyCode": "USD"
            },
            "VI": {
              "currencyCode": "USD"
            },
            "TF": {
              "currencyCode": "EUR"
            },
            "EH": {
              "currencyCode": "USD"
            },
            "BL": {
              "currencyCode": "EUR"
            }
        }
    }

}

module.exports = CONSTANTS;
